resolwe_runtime_utils.py

```diff
@@ -275,14 +301,233 @@ def checkrc(rc, *args):
+
+                if extracted_name.endswith('.tar'):
+                    with tarfile.open(extracted_name) as tar:
+                        tar.extractall()
+
+                    os.remove(extracted_name)
+                    extracted_name, _ = os.path.splitext(extracted_name)
+        else:
+            # Directory or several files in archive.
+            if imported_format != ImportedFormat.EXTRACTED:  # Compressed file required
+                with tarfile.open(extracted_name + '.tar.gz', 'w:gz') as tar:
+                    for fname in glob.glob(os.path.join(temp_dir, '*')):
+                        tar.add(fname, os.path.basename(fname))
+
+            if imported_format != ImportedFormat.COMPRESSED:  # Extracted file required
+                for path in os.listdir(temp_dir):
+                    shutil.move(os.path.join(temp_dir, path), './{}'.format(path))
+
+        shutil.rmtree(temp_dir)
+        return (
+            extracted_name + '.gz'
+            if imported_format == ImportedFormat.COMPRESSED
+            else extracted_name
+        )
+
+    def importUncompressed():
+        """Import uncompressed file."""
+        if imported_format != ImportedFormat.EXTRACTED:  # Compressed file required
+            with open(src, 'rb') as f_in, gzip.open(file_name + '.gz', 'wb') as f_out:
+                shutil.copyfileobj(f_in, f_out, CHUNK_SIZE)
+
+        if imported_format != ImportedFormat.COMPRESSED:  # Extracted file required
+            try:
+                shutil.copyfile(src, file_name)
+            except shutil.SameFileError:
+                pass  # Skip copy of downloaded files
+
+        return (
+            file_name + '.gz'
+            if imported_format == ImportedFormat.COMPRESSED
+            else file_name
+        )
+
+    # Large file download from Google Drive requires cookie and token.
+    try:
+        response = None
+        if re.match(
+            r'^https://drive.google.com/[-A-Za-z0-9\+&@#/%?=~_|!:,.;]*[-A-Za-z0-9\+&@#/%=~_|]$',
+            src,
+        ):
+            session = requests.Session()
+            response = session.get(src, stream=True)
+
+            token = None
+            for key, value in response.cookies.items():
+                if key.startswith('download_warning'):
+                    token = value
+                    break
+
+            if token is not None:
+                params = {'confirm': token}
+                response = session.get(src, params=params, stream=True)
+
+        elif re.match(
+            r'^(https?|ftp)://[-A-Za-z0-9\+&@#/%?=~_|!:,.;]*[-A-Za-z0-9\+&@#/%=~_|]$',
+            src,
+        ):
+            response = requests.get(src, stream=True)
+    except requests.exceptions.ConnectionError:
+        raise requests.exceptions.ConnectionError("could not connect to {}".format(src))
+
+    if response:
+        with open(file_name, 'wb') as f:
+            total = response.headers.get('content-length')
```

Comment:
that's annoying

Reply:
Do they return something else? Does the browser show progress for responses from Google Drive?

Reply:
Let me rephrase that. Responses for download of **large files** from Google Drive don't return content length. Large files are the ones that require confirming the download warning. In this case even the browser doesn't display download progress and the following headers are returned:
```
{
    'X-GUploader-UploadID':'AEnB2Ur7WZU4C0NuP7mtzw0mRVmdxm54cL2ksZE_35riixQNBk0bUJh9_6PZbWYbd2OB7NmDEZyH2Ku3oybkRdpmtZ7uU6MKQw',
    'Access-Control-Allow-Origin':'*',
    'Access-Control-Allow-Credentials':'false',
    'Access-Control-Allow-Headers':'Accept, Accept-Language, Authorization, Cache-Control, Content-Disposition, Content-Encoding, Content-Language, Content-Length, Content-MD5, Content-Range, Content-Type, Date, GData-Version, google-cloud-resource-prefix, Host, If-Match, If-Modified-Since, If-None-Match, If-Unmodified-Since, Origin, OriginToken, Pragma, Range, Slug, Transfer-Encoding, Want-Digest, x-chrome-connected, X-ClientDetails, X-Client-Version, X-Firebase-Locale, X-GData-Client, X-GData-Key, X-Goog-AdX-Buyer-Impersonation, X-Goog-Api-Client, X-Goog-AuthUser, X-Goog-PageId, X-Goog-Encode-Response-If-Executable, X-Goog-Correlation-Id, X-Goog-Request-Info, X-Goog-Request-Reason, X-Goog-Experiments, x-goog-iam-authority-selector, x-goog-iam-authorization-token, X-Goog-Spatula, X-Goog-Upload-Command, X-Goog-Upload-Content-Disposition, X-Goog-Upload-Content-Length, X-Goog-Upload-Content-Type, X-Goog-Upload-File-Name, X-Goog-Upload-Header-Content-Length, X-Goog-Upload-Offset, X-Goog-Upload-Protocol, X-Goog-Visitor-Id, X-Goog-FieldMask, X-HTTP-Method-Override, X-JavaScript-User-Agent, X-Pan-Versionid, X-Proxied-User-IP, X-Origin, X-Referer, X-Requested-With, X-Upload-Content-Length, X-Upload-Content-Type, X-Use-HTTP-Status-Code-Override, X-Ios-Bundle-Identifier, X-Android-Package, X-Ariane-Xsrf-Token, X-YouTube-VVT, X-YouTube-Page-CL, X-YouTube-Page-Timestamp, X-Goog-Meeting-Botguardid, X-Goog-Meeting-Debugid, X-Goog-Meeting-Token, X-Client-Data, X-Sfdc-Authorization, MIME-Version, Content-Transfer-Encoding',
    'Access-Control-Allow-Methods':'GET,OPTIONS',
    'Content-Type':'application/x-rpm',
    'Content-Disposition':'attachment;filename="atom.x86_64.rpm";filename*=UTF-8\'\'atom.x86_64.rpm',
    'Date':'Tue, 09 Apr 2019 10:07:44 GMT',
    'Expires':'Tue, 09 Apr 2019 10:07:44 GMT',
    'Cache-Control':'private, max-age=0',
    'X-Goog-Hash':'crc32c=QNomWA==',
    'Transfer-Encoding':'chunked',
    'Server':'UploadServer',
    'Alt-Svc':'quic=":443"; ma=2592000; v="46,44,43,39"'
}
```
Reponses for download of **small files** from Google Drive do have `Content-Length` header and download progress updates normally.

Reply:
> In this case even the browser doesn't display download progress

That's annoying.